### ISSUE TITLE
Attempted shadow->master transition without metadata - aborting

### ISSUE THREAD

**Issue by guestisp** [lizardfs/lizardfs#623]
When promoting a (new) shadow to master, i'm hitting the error:

```
Attempted shadow->master transition without metadata - aborting
```

isn't possible to promote a brand new server as master?
Running it directly as master works, but If I start as shadow and then trying to became a master, it fails.

**Comment by 4Dolio** [comment]
`lizardfs-admin promote-shadow 127.0.0.1 9419`

**Comment by guestisp** [comment]
The error is coming exactly running the command you wrote

**Comment by 4Dolio** [comment]
Can you confirm that the shadow is synced to the master? The cgi can tell you and a lizardfs-admin can also poll for that. The shadow should report the same metadata version as the master..

**Comment by guestisp** [comment]
Is not synced, there is nothing to sync, this is a brand new server thus master han not wrote anything this is why I've opened this issue

**Comment by 4Dolio** [comment]
Well.. That doesn't sound like an issue or bug to me. Even a virgin master at metedata ver 0 or 1 is still required for a shadow to sync to something I would expect.  Shadows are not intended to init from or promote from null.  Why do you expect that they would? What are you attempting to test exactly?

I suspect that even some metadata and CS comms get tacked onto the metadata stream. So the act of promotion of a shadow may well add a rev to the metadata for "master is exiting" and then perhaps "shadow got promoted" along with things like "chunk server blah attached" or "chunk server registered billion chunks".  I could be wrong on this last bit, but as I recall even an empty test cluster doing nothing increments metadata during transitions.

**Comment by guestisp** [comment]
I'm trying to set a scale-out master with automatic failover (I'm almost done) but I need to start an undefined number of metadata servers ready to be promoted

The only way to do this is starting in shasow

**Comment by 4Dolio** [comment]
Sure sure, my corosync HA can support an arbitrary number of shadows as well.  But from a cold start you will need to seed the cluster with one trusted master.  To accomplish this I insure that only a cleanly stopped singular master (no shadows present) is capable of seeding without intervention. Anything else is considered dirty and unable to init the cluster.

In my experience this needs to be init-personality of master and can not be done with init-personality of shadow followed by a promotion.  Because a shadow is only ready for promotion After it has synced up with a working master.

**Comment by psarna** [comment]
True. It's a precaution mechanism to avoid promoting an empty master by mistake. If you really want that, just start your master with initial-personality=master or let it synchronize first.

**Comment by guestisp** [comment]
> It's a precaution mechanism to avoid promoting an empty master by mistake

This seems reasonable.

So, when trying to promote an out-of-sync shadow to master, the operation fails and the out-of-sync shadow will stay as shadow. When another master became online, this out-of-sync shadow start to sync automatically ?

**Comment by psarna** [comment]
Yes, out-of-sync shadows continuously try to synchronize to master.

**Comment by guestisp** [comment]
What happens if this out-of-sync shadow, during it's sync phase, will be promoted as master?
Let's assume a very very huge master failing. the newly added shadow needs some time to be fully-synced (due to the amount of metadata to sync). What happens if this shadow will be promoted as master?

**Comment by 4Dolio** [comment]
Now that is a good question... i believe a not-yet-fully synced shadow should be considered not valid for promotion.  I am not sure if the master would honor a quick stop in this case? My corosync tracks the metadata rev of all members and would not promote a non-synced shadow.  But i am not positive how the native shadow mechanics might behave during this edge case.  So that is a good question, but I wont be testing it in my hugh production cluster..

**Comment by guestisp** [comment]
I can test that on my test cluster, when the orchestrator start to work. Adding a new key in Consul is easy, and I can use that for skipping out-of-sync cluster from leader promotion.

But.... how can I get the metadata version ?

**Comment by 4Dolio** [comment]
Probe with https://github.com/4Dolio/lizardfs/blob/4Dolio-corosync-ocf-metadataserver-patch-1/src/ha-cluster/metadataserver.in#L485 With the command `lizardfs-admin metadataserver-status --porcelain "${host}" "$matocl_port"` where host is always localhost.  Then we pull the rev from those results here https://github.com/4Dolio/lizardfs/blob/4Dolio-corosync-ocf-metadataserver-patch-1/src/ha-cluster/metadataserver.in#L640  within this lizardfs_master_monitor function.

**Comment by 4Dolio** [comment]
My metadataserver OCF script (https://github.com/4Dolio/lizardfs/blob/4Dolio-corosync-ocf-metadataserver-patch-1/src/ha-cluster/metadataserver.in) is admittedly not very pretty.  But I spent a good month working on it and encountering all sorts of edge cases and exceptions and attempted to handle them all as best I could.  I have been using that system for perhaps going on two years now to orchestrate about a dozen clusters with a variety of geometries and varying client loads.  It may not be perfect and surely has some flaws, but it gets the job done and has proven to be very reliable.  I can execute transitions between masters at will and under heavy client loads without any worry about disrupting clients.  If a cluster of masters is properly stopped (shadows first, then lone master) then it can also be started without any interventions.  I also took great care to ensure that it will never accidentally use stale metadata or seed the cluster backwards in time or to an empty state.  It will also attempt to spit out some helpful tips when it is unable to safely proceed without intervention.  As painful as it may be to try and read my ugly shell(bash) script, I left plenty of comments regarding why various tests are done in order to properly handle every possible type of unexpected event that I have ever encountered.  There is a wealth of accumulated knowledge in that little script, even if you use none of the script itself, or you (with good reason) despise my coding style...  While building your own HA process using whatever solution or language you choose, you may be able to learn some things or become aware of situations you may encounter along the way.  I am happy to answer any questions that arise due to reading my code and the resultant brain melting side effects.  ;)

**Comment by guestisp** [comment]
I've seen your script and I'm trying (as first step) to make the same but with Consul and not with corosync and peacemaker

I'm really close to have something finished and working where the whole stack is managed by Consul (using raft protocol). Our current script is a POC and still require lizardfs-admin and php-cli (much better than corosync and peacemaker)

My first step is to remove the lizardfs-admin requirement, the second step is to convert everything in go (a single executable that doesn't need installation or any external dependency, just copy it somewhere and you are ready) and as third step is to integrate the same in Lizard, totally removing the external requirements, making HA in Lizard native and based con raft protocol (because the official HA support, that we all waiting from years, will never be released, as it's a commercial product, exactly like in MooseFS)

**Comment by guestisp** [comment]
So, I'm hoping the publishing a PR adding an official HA support in Lizard will be accepted
There is no reason for an OSS to not accepted a PR fixing the only real missing thing wanted by everyone. The only reason to not accept that is for commercial purposes where Lizard want to sell it's own software thus adding the same as OSS won't be a great commercial move

**Comment by guestisp** [comment]
Anyway, 4Dolio, can you tell me exact steps on how to build your HA solution? I would like to test it

**Comment by 4Dolio** [comment]
https://github.com/4Dolio/lizardfs/tree/4Dolio-corosync-ocf-metadataserver-patch-1/src/ha-cluster The Readme describes two additional scripts.. One will 'fix' corosync and pacemaker on debian so they work for lfs and the metadataserver OCS script.  The other will guide the instantiation of the first corosync member.  Also keep in mind it was written for lfs 2.6.0 and as such you may need to symlink /etc/mfs to /etc/lizardfs for newer versions of lfs in the 3.x branch.. But I have used the same process with at least lfs some of the early 3.x versions.  The old lfs tools I also symlink to preserve their old command names, but I'm not sure that is needed for masters...  I also have a ton of .bashrc aliases and functions to aid it management operations which I have yet to share because of private information they might have.  You should enable the debugging features that I added to my script (and mfsmaster.conf to enable them)..  Then you can tail the logs and more easily watch what the cluster is doing.  If you get stuck I can try and answer more specific questions... And I will try to sanitize and pass along some of my other OPs tools.  But for tonight, I really need to turn in...

**Comment by 4Dolio** [comment]
It should work with deb 6 and 7..

**Comment by guestisp** [comment]
There is a CMakeList, shoul I run "cmake" to build something or the only required steps are what is described in readme ?

**Comment by 4Dolio** [comment]
Just the readme... Ignore the CMakeList..  The whole thing is a branch from an original HA-Wizard tool which I was never able to get to work properly... But the metadataserver(.in) is the main magic.. the other .sh mentioned in the readme just prep and init the corosync+pacemaker services.. The metadataserver.in gets the .in removed and dropped into wherever that script would normally lives .... somewhere near /usr/lib/ocf/*/lizardfs I think...  Of course there may be things that I neglect to mention because they are obvious (to me) or some other part of my internal documentation... So if you get stuck, I would like to fix my readme and/or scripts...

**Comment by guestisp** [comment]
If doesn't require too much time from you, could you please share which checks are you making to see if local shadow can be safely promoted as master ?

I've seen that there are multiple checks in multiple points. If I understood properly, currenty you are checking that local shadow has some metadata (avoiding to promote an empty shadow) and local shadow as metadata revision >= the latest known master medatarevisio (thus, only "synced" shadow can be promoted)

Probably, a list of check would be great. If you don't have time, don't worry, i'll try to figure out on my own.

**Comment by 4Dolio** [comment]
That sounds about right.. shadow must respond that it is shadow and rev must be >= to the cluster(corosync) recorded metadata rev (as updated by the master)... Master must completely exit before shadow is chosen, IP is then moved and shadow is promoted.  Shadow will always follow the stopping master up to the most current rev.  If no shadows the master will refuse to quick-stop and so will fall back to normal/slow stop including writing metadata to disk.  The original script author did most of the logic work (and good clean code).. I just hacked it up so that it would actually function reliably and handle various exceptions.

**Comment by guestisp** [comment]
So, you always try to "kill" master in fast way (`lizardfs-admin stop-master-without-saving-metadata`), if shadows are connected, the operation will succeeed, else it stop slowly.

**Comment by 4Dolio** [comment]
Indeed.

**Comment by 4Dolio** [comment]
But it is not a literal "kill".

**Comment by 4Dolio** [comment]
I takes a master with 44GiB of metadata many minutes to exit.. and you can not promote a shadow untill the old master completely exits. So it must exit as quickly as possible.  Thus the lizardfs-admin ability to ask the master so do something that is otherwise insane and dangerous, to exit without saving metadata to disk.. and volla

**Comment by guestisp** [comment]
Ok. The `shadow/syncing` status is made by your script or is something returned by `lizardfs-admin` ?

**Comment by 4Dolio** [comment]
The corosync metadataserver bash script calls the lizardfs-admin to poll and prompt the master service.
At any moment (at 1 sec intervals for the master, and 2 second intervals for shadows) the script is called, and then itself calls lizardfs-admin to collect states and revisions.  I do not know specifically what is collected and parsed and used to decide things... That is why I spent a month hacking the closest thing to a core HA solution, the original corosync wizard that never worked for me.  Each master is aware of it's rev, and the master records it to corosync ( the cluster, all members, I also did not write corosync or ask for them to fix stuff ).  I just tweaked it until it worked reliably enough for me to trust it with my data and job.  It is not perfect and faults happen but I have not lost anything yet, praise to the bit gods.

I use lizardfs-admin to do many but not all of the things that lizardfs-admin is capable of doing.  It is what enables corosync and my hacked up sloppy script to pull off HA, homebrew style.  Others have had success with things like CARP, which I also used to use.  Or other HA solutions, or you can buy it from the team.  Or you can reference direct lines in my script and I will answer specific questions.  As for lizardfs-admin, it is simply an extension to the master service to allow for interesting, dangerous, beautiful, and awesome interactions.

The shadow/syncing is just something that masters and shadows do.  I'm not even sure you can use those ha-managed-personalities, without the lizardfs-admin tool though...

capisci?

**Comment by guestisp** [comment]
@psarna Trying to change the master IP (all shadows are connected to `mfsmaster` hosts resolved by `/etc/hosts.conf`), i'm getting this:

```
Nov 15 16:42:46 x mfsmaster[18613]: connecting to Master
Nov 15 16:42:46 x mfsmaster[18613]: connected to Master
Nov 15 16:42:46 x mfsmaster[18613]: Cannot register to master: Incorrect register BLOB
Nov 15 16:42:46 x mfsmaster[18613]: main master server module: got invalid message in shadow state (type:400)
Nov 15 16:42:46 x mfsmaster[18613]: main master server module: got invalid message in shadow state (type:400)
```

What does it mean ? The master is running properly:

```
# lizardfs-admin metadataserver-status localhost 9421
     personality: master
   server status: running
metadata version: 1513
```

but this shadow is stuck in disconnected mode:
```
# lizardfs-admin metadataserver-status localhost 9421
     personality: shadow
   server status: disconnected
metadata version: 0
```

**Comment by psarna** [comment]
It means a version mismatch, i.e. your shadow is in newer version than master. Due to our backwards compatibility rules, shadows can have higher version than master, but not the other way round.

**Comment by guestisp** [comment]
> shadows can have higher version than master, but not the other way round.

So , is the opposite than 

> your shadow is in newer version than master

My shadow is older than master?

I'll check.

Can I suggest a more easier error message?
Is possible to temporarily remove this check?
